pkg/debuginfo/manager.go

```diff
@@ -226,26 +230,34 @@ func (di *Manager) ensureUploaded(ctx context.Context, objFile *objectfile.Mappe
 		r = f
 	}
 
-	hash, err := hash.Reader(r)
-	if err != nil {
-		return fmt.Errorf("hash debuginfos: %w", err)
-	}
+	// The hash is cached to avoid re-hashing the same binary
+	// and getting to the same result again.
+	var (
+		h   string
+		err error
+	)
+	if v, ok := di.hashCache.Load(buildID); ok {
+		h = v.(string) //nolint:forcetypeassert
+	} else {
+		h, err = hash.Reader(r)
+		if err != nil {
+			return fmt.Errorf("hash debuginfos: %w", err)
+		}
+		di.hashCache.Store(buildID, h)
 
-	if _, err := r.Seek(0, io.SeekStart); err != nil {
-		return fmt.Errorf("failed to seek to the beginning of the file: %w", err)
+		if _, err = r.Seek(0, io.SeekStart); err != nil {
+			return fmt.Errorf("failed to seek to the beginning of the file: %w", err)
+		}
 	}
 
 	initiateResp, err := di.debuginfoClient.InitiateUpload(ctx, &debuginfopb.InitiateUploadRequest{
 		BuildId: buildID,
-		Hash:    hash,
+		Hash:    h,
```

Comment:
I was thinking to switch to highwayhash, but I wasn't sure if there is a specific reason to stick to xxhash in debuginfo package, i.e., xxhash returns strings such as `2402cb2193253c5c`, and highwayhash returns uint64.

Note, highwayhash has been used in the following files.

```
./pkg/ksym/ksym.go:	"github.com/parca-dev/parca-agent/pkg/hash"
./pkg/perf/perf.go:	"github.com/parca-dev/parca-agent/pkg/hash"
./pkg/process/maps.go:	"github.com/parca-dev/parca-agent/pkg/hash"
```